source/common/upstream/cluster_manager_impl.cc

```diff
@@ -960,8 +965,13 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(ClusterManagerCluster& cm_
         }
 
         if (new_cluster != nullptr) {
-          for (auto& cb : cluster_manager->update_callbacks_) {
-            cb->onClusterAddOrUpdate(*new_cluster);
+          // The following iteration calls the onClusterAddOrUpdate callback function,
+          // while allowing the callbacks to remove themselves.
+          auto it = cluster_manager->update_callbacks_.begin();
+          while (it != cluster_manager->update_callbacks_.end()) {
+            auto next = std::next(it);
+            (*it)->onClusterAddOrUpdate(*new_cluster);
+            it = next;
```

Comment:
It may be worth covering the case of additions to update_callbacks_ while iterating.  I think those callbacks would execute if they are added by a callback that is not the last one in the list.  A possibly better behavior would be to ASSERT that there are no additions while iterating.

I'm of course assuming that allowing removals during iteration is desirable.  We could decide that we do not want to allow additions or removals while iterating.